File: day1.jl
using Test
using CSV

function fuelFor(mass)
    return convert(Int64, floor(mass/3) - 2)
end

function totalFuelFor(mass)
    fuel = fuelFor(mass)
    if fuel <= 0
        return 0
    end

    return fuel + totalFuelFor(fuel)
end

function test()
    @test fuelFor(12) == 2
    @test fuelFor(14) == 2
    @test fuelFor(1969) == 654
    @test fuelFor(100756) == 33583
    @test totalFuelFor(14) == 2
    @test totalFuelFor(1969) == 966
    @test totalFuelFor(100756) == 50346
end

function part1(mass)
    fuel = sum(map(fuelFor, mass))
    println("part1: $fuel")
end        

function part2(mass)
    totalFuel = sum(map(totalFuelFor, mass))
    println("part2: $totalFuel")
end

test()

data = CSV.read(joinpath("inputs", "day1.txt"), datarow=1)
mass = data[:, 1]
part1(mass)
part2(mass)

File: day1.py
import csv
import math
import os

import pytest


def fuelFor(mass: int) -> int:
    return int(math.floor(mass/3) - 2)


def totalFuelFor(mass: int) -> int:
    fuel = fuelFor(mass)
    if fuel <= 0:
        return 0
    
    return fuel + totalFuelFor(fuel)

@pytest.mark.parametrize("input, output", [
    (12, 2),
    (14, 2),
    (1969, 654),
    (100756, 33583)
])
def test_fuelFor(input, output):
    assert fuelFor(input) == output


@pytest.mark.parametrize("input, output", [
    (14, 2),
    (1969, 966),
    (100756, 50346)
])
def test_totalFuelFor(input, output):
    assert totalFuelFor(input) == output


def part1(mass):
    fuel = sum([fuelFor(m) for m in mass])
    print("part1:", fuel)


def part2(mass):
    fuel = sum([totalFuelFor(m) for m in mass])
    print("part2:", fuel)


def _main():
    with open(os.path.join("inputs", "day1.txt")) as file:
        mass = [int(line) for line in file]

    part1(mass)
    part2(mass)


if __name__ == "__main__":
    _main()
